error-curation.md

```diff
@@ -0,0 +1,87 @@
+# Error Curation
+
+## Overview
+
+Many errors occur in Temporal yet pithy error messages alone are not enough to give context. We need to provide
+easy-to-reference identifiers to each and pages on each to help.
+
+**Goals**
+
+These are goals for the error curation project, not this proposal in particular though there is overlap.
+
+* Document requirements for rule numbering/identification
+* Document requirements for rule documenting
+* Provide home for all rule documents
+* Build initial set of rules for use in SDKs
+
+**Non-Goals**
+
+* Document every error
+* Make the error documents available on docs.temporal.io at first
+* Provide any UX linking errors from UI, CLI, etc
+* Extremely strict documentation guidelines
+
+## Approach
+
+This approach is inspired by .NET analyzers which have coded values and dedicated pages for each error. For example,
+see:
+
+* https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/categories (e.g.
+  https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1001)
+* https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/DOCUMENTATION.md (e.g.
+  https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1200.md)
+* https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/index.md (e.g.
+  https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD003.md)
+
+Solution:
+
+* A "rule" is referenceable, documented error/warning/note/guideline
+* We are calling these "rules" not "errors"
+  * They may be found eagerly, may just be disable-able warnings, or may even be just best practice rules
+* Rules can be for many things:
+  * Errors (e.g. SDK errors, static analysis errors, server errors, etc)
+  * Guidelines (e.g. best practices, warnings, etc)
+* Every rule has an identifier in the form of `TMPRLXXXX`
+  * "Categories" of rules are given a 100 block as rarely is more needed
+    * Category 100 blocks should try not to be next to another so space can be given in some rare case we exceed 100 in
+      a category
+    * There are no rules on what a category is, just a grouping of like things
+  * The rules within a category do not have to be ordered in any manner. Usually you just use the next available.
+  * `0` is never used in the first two digits
+  * Can discuss the first thousands-block digit though initial thought for first two is:
+    * `TMRPL1XXX` block is for SDK-specific and workflow/activity authoring things
+    * `TMPRL2XXX` block is for server-specific and operation things
+* Rule/error messages across Temporal should be prepended with `"[TMPRLXXXX] "` (sans quotes)
```

Comment:
For implementing this in Go, I think we can leverage golangci-lint, which has things like enforced error-wrapping via goerr113

Reply:
I think that's overthinking at first, but if that's what y'all want to do in your codebase, ok. Not only would I not expect these identifiers on every error (that's silly, not all errors deserve a rule), there's no need to add friction. I've seen these projects get out of hand with overthink/overabstraction or too-rigid of requirements on developers. Just keep it simple, prepend your error string literals on errors that deserve to be cataloged, and move on.

Reply:
How (if at all) are we going to try to maintain a correspondence between these error codes and (future) error codes in our source?  A small group of us are considering improvements to how we code error logging, and one popular idea is to assign unique error codes to each log message.  Many of these code locations are also the “source” of the errors you are delineating in this proposal. Maybe we should just admit that trying for any synchrony here is a fool’s game, but I wanted to at least raise the question.  Also note that there will be a lot more than 1000 of the other error codes. I wonder if you shouldn’t add one more digit just on general principles

Reply:
This is one of the reasons I avoided calling these error codes. They are more "rules" and only applicable for a reasonable subset of situations worth documenting. Some deep down software error may not really be worth having its own page with mitigation and everything. And sometimes many errors are really violations of the same rule (e.g. `TMPRL1322` may be for invalid input/output data type and put on lots of errors in many SDKs).

The goal here is more to have a reasonably sized set of identifiers that users can reference and read about. If there is an effort to add error codes at a lower level or more granular, I think it should be a bit different than these identifiers.

> Maybe we should just admit that trying for any synchrony here is a fool’s game

Yes! This is exactly where I'm at. If we keep a reasonably sized enough list, not only can a developer go see if a situation applies to the error they are about to throw, but a user can literally scroll the list to see bad things they could do. That's why I linked the .NET rules as inspiration because they fit more of my mental model of rules vs error codes.

I have no doubt we'll miss some, have to go add later, etc. But I am hoping we don't get too granular with them.

> Also note that there will be a lot more than 1000 of the other error codes. I wonder if you shouldn’t add one more digit just on general principles

8099 in this case I think or something like that (`1100-9999` with no `0`s in second digit), but really for me it was about identifiers vs string length and the category blocks. For how I was hoping they'd be used, I didn't figure a category would need more than 100. Honestly don't even care that much about first digit, I have ~80 categories to give out which I think is ok.

If we do thing we need more, I would support alternative prefixes besides `TMPRL` which is already a mouthful. And maybe decimals, e.g. `TMPRL3563.12` but I am hoping not tbh. Or we could even get into letters for digit parts. I'd be willing to make this 5 digits if we can take a character off `TMPRL` somewhere.

Reply:
TMPRL => TIO? (aka t.io)
I keep hoping we'll actually buy that domain!

Reply:
We do own `t.mp` and I fully expect something like `t.mp/rules/TMPRL1234` does the proper redirection. Also unsure if `.io` is something we want to make part of the company abbreviation name here. `TPRL` is an option. Will let others weigh in.

Reply:
I do think every genuinely _error_-level event that becomes user visible should get one of these. However, finding those situations statically isn't suuuuper doable.

Probably, the best approach here is to be able to run the linter and then just mark a bunch of stuff as ignored. I think it'd be good to have that turned on, so that devs adding new errors are forced to choose between adding the code and saying "no it's ok this gets an ignore". It's a slight annoyance but one that will ensure we keep up with this effort long run.

Reply:
I like having identifiers that are somewhat self-describing (e.g. `[TMPRL.sdk.whatever]`) but also agree with the need for brevity. Just noting it here as another option should we ever find ourselves with too many of these things (which I agree is unlikely—more than even ~1,000 rules means we probably need to rethink our system >.>).

Reply:
Not just brevity, but people are bad at naming those identifiers IME